Add more info to existing samples (#46660)
Azure/azure-content: articles/stream-analytics/stream-analytics-parsing-json.md

    DeviceID,
    Location.Lat,
    Location.Long,
    SensorReadings.Temperature,
    SensorReadings.SensorMetadata.Version
FROM input
```

The result is:

|DeviceID|Lat|Long|Temperature|Version|
|-|-|-|-|-|
|12345|47|122|80|1.2.45|

### Select all properties
You can select all the properties of a nested record using '*' wildcard. Consider the following example:

```SQL
SELECT
    DeviceID,
    Location.*
FROM input
```

The result is:

|DeviceID|Lat|Long|
|-|-|-|
|12345|47|122|

### Access nested fields when property name is a variable

Use the [GetRecordPropertyValue](https://docs.microsoft.com/stream-analytics-query/getmetadatapropertyvalue) function if the property name is a variable. This allows for building dynamic queries without hardcoding property names.

For example, imagine the sample data stream needs **to be joined with reference data** containing thresholds for each device sensor. A snippet of such reference data is shown below.

```json
{
    "DeviceId" : "12345",
    "SensorName" : "Temperature",
    "Value" : 85
},
{
    "DeviceId" : "12345",
    "SensorName" : "Humidity",
    "Value" : 65
}
```

> The goal here is to join our sample dataset from the top of the article to that reference data, and output one event for each sensor measure above its threshold. That means our single event above can generate multiple output events if multiple sensors are above their respective thresholds, thanks to the join. To achieve similar results without a join, see the chapter below.

Reviewer: ```suggestion
The goal here is to join our sample dataset from the top of the article to that reference data, and output one event for each sensor measure above its threshold. That means our single event above can generate multiple output events if multiple sensors are above their respective thresholds, thanks to the join. To achieve similar results without a join, see the section below.
```

```SQL
SELECT
    input.DeviceID,
    thresholds.SensorName,
    "Alert : Sensor above threshold" AS AlertMessage
FROM input      -- stream input
JOIN thresholds -- reference data input
ON
    input.DeviceId = thresholds.DeviceId
WHERE
    GetRecordPropertyValue(input.SensorReadings, thresholds.SensorName) > thresholds.Value
```

> The **GetRecordPropertyValue** selects the property in SensorReadings which name matches the property name coming from the reference data, then it extracts the value associated from SensorReadings.

Reviewer: ```suggestion
**GetRecordPropertyValue** selects the property in *SensorReadings*, which name matches the property name coming from the reference data. Then the associated value from *SensorReadings* is extracted.
```

CROSS APPLY GetRecordProperties(event.SensorReadings) AS sensorReading
```

The result is:

|DeviceID|SensorName|AlertMessage|
|-|-|-|
|12345|Temperature|80|
|12345|Humidity|70|
|12345|CustomSensor01|5|
|12345|CustomSensor02|99|
|12345|SensorMetadata|[object Object]|

> Using [WITH](https://docs.microsoft.com/en-us/stream-analytics-query/with-azure-stream-analytics), it's then possible to route those events to different destinations:

Reviewer 2: Please remove `en-us` from all links. The system will pick the locale according to the user's preferences.

```SQL
WITH Stage0 AS
(
    SELECT
        event.DeviceID,
        sensorReading.PropertyName,
        sensorReading.PropertyValue
    FROM input as event
    CROSS APPLY GetRecordProperties(event.SensorReadings) AS sensorReading
)

SELECT DeviceID, PropertyValue AS Temperature INTO TemperatureOutput FROM Stage0 WHERE PropertyName = 'Temperature'
SELECT DeviceID, PropertyValue AS Humidity INTO HumidityOutput FROM Stage0 WHERE PropertyName = 'Humidity'
```

## Array data types

Array data types are an ordered collection of values. Some typical operations on array values are detailed below. These examples use the functions [GetArrayElement](https://docs.microsoft.com/stream-analytics-query/getarrayelement-azure-stream-analytics), [GetArrayElements](https://docs.microsoft.com/stream-analytics-query/getarrayelements-azure-stream-analytics), [GetArrayLength](https://docs.microsoft.com/stream-analytics-query/getarraylength-azure-stream-analytics), and the [APPLY](https://docs.microsoft.com/stream-analytics-query/apply-azure-stream-analytics) operator.

> Here is example of a single event. Both `CustomSensor03` and `SensorMetadata` are of type **array**:

Reviewer: ```suggestion
Here is an example of a single event. Both `CustomSensor03` and `SensorMetadata` are of type **array**:
```

|-|
|12|

### Select array length

```SQL
SELECT
    GetArrayLength(SensorReadings.CustomSensor03) AS arrayLength
FROM input
```

The result is:

|arrayLength|
|-|
|3|

### Convert array elements into separate events

Select all array element as individual events. The [APPLY](https://docs.microsoft.com/stream-analytics-query/apply-azure-stream-analytics) operator together with the [GetArrayElements](https://docs.microsoft.com/stream-analytics-query/getarrayelements-azure-stream-analytics) built-in function extracts all array elements as individual events:

```SQL
SELECT
    DeviceId,
	CustomSensor03Record.ArrayIndex,
	CustomSensor03Record.ArrayValue
FROM input
CROSS APPLY GetArrayElements(SensorReadings.CustomSensor03) AS CustomSensor03Record

```

The result is:

|DeviceId|ArrayIndex|ArrayValue|
|-|-|-|
|12345|0|12|
|12345|1|-5|
|12345|2|0|

```SQL
SELECT   
    i.DeviceId,	
    SensorMetadataRecords.ArrayValue.smKey as smKey,
    SensorMetadataRecords.ArrayValue.smValue as smValue
FROM input i
CROSS APPLY GetArrayElements(SensorMetadata) AS SensorMetadataRecords
 ```
 
The result is:

|DeviceId|smKey|smValue|
|-|-|-|
|12345|Manufacturer|ABC|
|12345|Version|1.2.45|

> Using [WITH](https://docs.microsoft.com/en-us/stream-analytics-query/with-azure-stream-analytics), if necessary it's then possible to [join](https://docs.microsoft.com/en-us/stream-analytics-query/join-azure-stream-analytics) those fields in the main query (since this is not a join on reference data, it needs a [time boundary](https://docs.microsoft.com/en-us/stream-analytics-query/join-azure-stream-analytics#BKMK_DateDiff) that prevents duplication).

Reviewer: This line is a little confusing. Can you please clarify?

Author: It is confusing indeed :)

The use case here is wanting to have Manufacturer and Version in columns instead of row. 
You will need to pivot the dataset in order to achieve that. 

One approach to pivoting is to join the processed dataset (where you have extracted your values in rows) with the original one (with the right cardinality of 1 row per incoming event). 

To do that in ASA you'll need to use 2 syntax elements: 

- the WITH clause, because you need to join the processed dataset with the original one as many times as columns you want to extract
- the JOIN clause. For that JOIN, you will need to define a time boundary in the join conditions (`DATEDIFF(...) BETWEEN ...`) because it's joining 2 live datasets. I wanted to make a note of that because it's not the case in the example give a bit earlier that was joining a reference dataset

Reviewer 2: @field Thanks for contributing to Docs! Please remove `en-us` from the 3 links. Then those will work in all locales. ~Jason

Author: Rephrased in commit [
5682c17](https://github.com/MicrosoftDocs/azure-docs/pull/46660/commits/5682c17eb5b853e829f987b2a36c6e8146c63da9) below
